src/binder/query/query_graph_label_analyzer.cpp

```diff
@@ -94,49 +95,231 @@ void QueryGraphLabelAnalyzer::pruneNode(const QueryGraph& graph, NodeExpression&
+        stepFromRightGraph[dstTableID][srcTableID].push_back(relEntry.getTableID());
+        if (relDirectionType == RelDirectionType::BOTH) {
+            stepFromLeftGraph[dstTableID][srcTableID].push_back(relEntry.getTableID());
+            stepFromRightGraph[srcTableID][dstTableID].push_back(relEntry.getTableID());
+        }
+    }
+    auto stepFromLeft =
+        pruneRecursiveRel(stepFromLeftGraph, srcTableIDSet, dstTableIDSet, lowerBound, upperBound);
+    auto stepFromRight =
+        pruneRecursiveRel(stepFromRightGraph, dstTableIDSet, srcTableIDSet, lowerBound, upperBound);
+    return {stepFromLeft, stepFromRight};
+}
+
+std::vector<table_id_set_t> QueryGraphLabelAnalyzer::pruneRecursiveRel(
+    const std::unordered_map<table_id_t, std::unordered_map<table_id_t, table_id_vector_t>>& graph,
+    const table_id_set_t& startTableIDSet, const table_id_set_t& endTableIDSet, size_t lowerBound,
+    size_t upperBound) const {
+
+    // There may be multiple rel type between src and dst
+    using PATH = std::vector<table_id_vector_t>;
+    std::vector<std::vector<PATH>> paths(upperBound + 1);
+
+    std::function<void(table_id_t, size_t, std::vector<table_id_vector_t>&)> dfs;
+    dfs = [&](table_id_t curTableID, size_t currentLength, std::vector<table_id_vector_t>& path) {
+        if (currentLength >= lowerBound && currentLength <= upperBound &&
+            endTableIDSet.contains(curTableID)) {
+            paths[currentLength].push_back(path);
+        }
+        if (currentLength >= upperBound) {
+            return;
+        }
+        if (graph.contains(curTableID)) {
+            for (const auto& [neighbor, relTableID] : graph.at(curTableID)) {
+                path.push_back(relTableID);
+                dfs(neighbor, currentLength + 1, path);
+                path.pop_back();
             }
         }
-    } else {
-        auto srcTableIDSet = rel.getSrcNode()->getTableIDsSet();
-        auto dstTableIDSet = rel.getDstNode()->getTableIDsSet();
-        for (auto& entry : rel.getEntries()) {
-            auto& relEntry = entry->constCast<RelTableCatalogEntry>();
-            auto srcTableID = relEntry.getSrcTableID();
-            auto dstTableID = relEntry.getDstTableID();
-            if (!srcTableIDSet.contains(srcTableID) || !dstTableIDSet.contains(dstTableID)) {
-                continue;
+    };
+    for (auto startTableID : startTableIDSet) {
+        std::vector<table_id_vector_t> path;
+        dfs(startTableID, 0, path);
+    }
+    // merge
+    std::vector<table_id_set_t> stepActiveTableIDs(upperBound);
+    for (auto i = lowerBound; i <= upperBound; ++i) {
+        for (auto path : paths[i]) {
+            for (auto j = 0u; j < i; ++j) {
+                auto rels = path[j];
+                stepActiveTableIDs[j].insert(rels.begin(), rels.end());
             }
-            prunedEntries.push_back(entry);
         }
     }
-    rel.setEntries(prunedEntries);
+    return stepActiveTableIDs;
```

Comment:
Let me try to summarize the logic here and pls correct me if I'm wrong.

Consider topology `Person - knows -> Person` and `Person - livesIn -> City` and query 
```
MATCH (a:Person) - [*2..2] -> (b:City)
``` 
WLOG, we assume going from left to right.

The `dfs` part will first generate all size 2 path
```
<<knows>, <knows>>
<<knows>, <livesIn>>
```
and because `<<knows>, <knows>>` dst is not `b:City` it will be pruned.

So eventually we only compute `<<knows>, <livesIn>>`.

If my understanding is correct, can u also give a benchmark number in the PR description to make sure we actually get performance benefit.